components/brave_shields/browser/ad_block_engine.cc

```diff
@@ -253,49 +265,73 @@ void AdBlockBaseService::OnGetDATFileData(base::OnceClosure callback,
   std::move(callback).Run();
 }
 
-void AdBlockBaseService::UpdateAdBlockClient(
+void AdBlockEngine::UpdateAdBlockClient(
     std::unique_ptr<adblock::Engine> ad_block_client) {
   DCHECK(GetTaskRunner()->RunsTasksInCurrentSequence());
   ad_block_client_ = std::move(ad_block_client);
   AddKnownTagsToAdBlockInstance();
-  AddKnownResourcesToAdBlockInstance();
+  DemandResourceReload();
 }
 
-void AdBlockBaseService::AddKnownTagsToAdBlockInstance() {
+void AdBlockEngine::AddKnownTagsToAdBlockInstance() {
   std::for_each(tags_.begin(), tags_.end(),
                 [&](const std::string tag) { ad_block_client_->addTag(tag); });
 }
 
-void AdBlockBaseService::AddKnownResourcesToAdBlockInstance() {
-  ad_block_client_->addResources(resources_);
+void AdBlockEngine::OnListSourceLoaded(const DATFileDataBuffer& list_source) {
+  GetTaskRunner()->PostTask(
+      FROM_HERE, base::BindOnce(&AdBlockEngine::UpdateFiltersOnFileTaskRunner,
+                                base::Unretained(this), list_source));
 }
 
-bool AdBlockBaseService::Init() {
-  return true;
+void AdBlockEngine::UpdateFiltersOnFileTaskRunner(
+    const DATFileDataBuffer& filters) {
+  DCHECK(GetTaskRunner()->RunsTasksInCurrentSequence());
+
+  UpdateAdBlockClient(std::make_unique<adblock::Engine>(
+      reinterpret_cast<const char*>(filters.data()), filters.size(),
+      redirect_url_parsing_));
 }
 
-void AdBlockBaseService::ResetForTest(const std::string& rules,
-                                      const std::string& resources,
-                                      bool include_redirect_urls) {
-  DCHECK(GetTaskRunner()->RunsTasksInCurrentSequence());
-  // This is temporary until adblock-rust supports incrementally adding
-  // filter rules to an existing instance. At which point the hack below
-  // will dissapear.
-  ad_block_client_.reset(new adblock::Engine(rules, include_redirect_urls));
-  AddKnownTagsToAdBlockInstance();
-  if (!resources.empty()) {
-    resources_ = resources;
+void AdBlockEngine::OnDATLoaded(const DATFileDataBuffer& dat_buf) {
+  // An empty buffer will not load successfully.
+  if (dat_buf.empty()) {
+    return;
   }
-  AddKnownResourcesToAdBlockInstance();
+  GetTaskRunner()->PostTask(
+      FROM_HERE, base::BindOnce(&AdBlockEngine::UpdateDATOnFileTaskRunner,
+                                base::Unretained(this), dat_buf));
+}
+
+void AdBlockEngine::UpdateDATOnFileTaskRunner(
+    const DATFileDataBuffer& dat_buf) {
+  DCHECK(GetTaskRunner()->RunsTasksInCurrentSequence());
+  auto e = std::make_unique<adblock::Engine>();
+  e->deserialize(reinterpret_cast<const char*>(&dat_buf.front()),
+                 dat_buf.size());
+
+  UpdateAdBlockClient(std::move(e));
+}
+
+void AdBlockEngine::OnResourcesLoaded(const std::string& resources_json) {
+  ad_block_client_->addResources(resources_json);
 }
 
-///////////////////////////////////////////////////////////////////////////////
+void AdBlockEngine::DemandResourceReload() {
```

Comment:
this name is very odd to me, why isn't it just `LoadResources`?

Reply:
I named it that way because in the current implementation, there's no guarantee that the callback will be called with new resources (if the ResourceProvider still isn't ready yet).